### PR TITLE
SparkSQL: Support for querying snapshots when reading data with Delta Lake

### PR DESCRIPTION
**Pull request by R7L208**
### Brief summary of the change made

This added new segments and grammar, as well as updating `TableExpressionSegment` to support querying historical snapshots of Delta Tables

### Are there any other side effects of this change that we should be aware of?
- I renamed an instance of `Spark3` to `SparkSQL` in the README that was missed.
- update `atsign_literal` to `at_sign_literal` in BigQuery, Exasol, and MySQL

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by R7L208** [comment]
I noticed that several other dialects implement a variation of `at_sign_literal` and a lexer for `@`. Is it worthwhile to move this into ANSI and then update dialects that inherit from it for specific syntax?

**Comment by tunetheweb** [comment]
> Reference in new issue

Possibly, though it also means something else in other dialects (e.g. Oracle) so might not work. Willing to see.

However, you have a different definition here so could have a common one anyway:
- You allow starting with numbers, bigquery and exasol don't. Does sparksql allow numbers? Does it allow underscores like the others do? Or would you propose overriding it here to a slightly different definition? In which case not totally convinced it's worth it.
- You've called it `at_sign_literal`, while bigquery and exasol call it `atsign_literal`. I actually prefer yours but can we change the others for consistency? Or change yours if you prefer?

**Comment by R7L208** [comment]
> You allow starting with numbers, bigquery and exasol don't. MySQL allows anything. Does sparksql allow starting with numbers? Or would you propose overriding it here to a slightly different definition? In which case not totally convinced it's worth it.

The only two uses for an `@` and inherently the `at_sign_literal` in SparkSQL is to specify a historical snapshot of a Delta Table which would be one of `@v###` to specify by version number or `@yyyymmddHHSSss` where the timestamp is referenced numerically (ie `19911002123000`) to reference a version by timestamp. I kept it broad since it would correctly parse those two segments and still be syntactically valid if used in other cases within Spark (but pretty sure it's otherwise not used).

I'm also not convinced it's worth it. Sometimes it seems much more cut and dry but pieces of it could be moved to ANSI, like having a lexer for `@`, that would allow for simplifying segments in other dialects. Definitely provides questionable improvements which would be marginal at best.

> You've called it at_sign_literal, while bigquer, exasol and mysql call it atsign_literal. I actually prefer yours but can we change the others for consistency? Or change yours if you prefer?

I also prefer at_sign_literal and will make the update in other dialects. 🙂

**Comment by tunetheweb** [comment]
> like having a lexer for `@`, that would allow for simplifying segments in other dialects

I don't think we can do that even as we lex the whole literal and not just the at sign.

**Comment by R7L208** [comment]
> > like having a lexer for `@`, that would allow for simplifying segments in other dialects
> 
> I don't think we can do that even as we lex the whole literal and not just the at sign.

That is an excellent point. Seems like there's no compelling argument to move this into ANSI unless multiple at_sign_literals are defined, ie at_sign_numeric_literal and at_sign_alpha_literal, but I think that is unnecessary given current usage

**Comment by tunetheweb** [comment]
Agreed. Though good to look for things to have the discussion!

File: src/sqlfluff/dialects/dialect_sparksql.py [hunk 1]
```diff
@@ -386,6 +393,21 @@
         # there since there are no significant syntax changes
         "JDBC",
     ),
+    AsOfGrammar=Sequence("AS", "OF"),
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_sparksql.py:396]: Not sure this is really necessary. Only use it twice and the grammar shortcut is longer than than the actual grammar! That can make sense if want to override it in other dialects but not so sure here.

**Review by tunetheweb — CHANGES_REQUESTED** [review]
In general looks good. Think we should change the AsOfGrammar as noted. What do you think?

**Comment by R7L208** [comment]
> In general looks good. Think we should change the AsOfGrammar as noted. What do you think?

Agreed! I can make those changes early tomorrow

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**